### PR TITLE
IGAPP-1217: Cut loading times

### PR DESCRIPTION
**Pull request by steffenkleinle**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

### CODE REVIEW

**Comment by steffenkleinle** [comment]
This cuts the loading times by using the single city endpoint for city content routes (IGAPP-1267) and executing requests simultaneously (IGAPP-1217).

File: web/src/CityContentSwitcher.tsx
Comment (steffenkleinle) [inline web/src/CityContentSwitcher.tsx:63]: This is the important change that renders the routes and therefore loads the other requests without waiting for the loading of the city.

File: web/src/CityContentSwitcher.tsx
Comment (steffenkleinle) [inline web/src/CityContentSwitcher.tsx:105]: If the city is not available yet nothing is rendered in the routes anyway, therefore render the route until we know if its enabled.

**Comment by steffenkleinle** [comment]
@SteffiStuffel we can probably close the epic once this two PRs are merged right? https://issues.tuerantuer.org/browse/IGAPP-1216 Or anything else left to do?

File: web/src/CityContentSwitcher.tsx
Comment (sarahsporck) [inline web/src/CityContentSwitcher.tsx:134]: so we could also change the `CityRouteProps` to 
```ts
type CityRouteProps {
  city: CityModel
  pathname: string
  cityCode: string
  languageCode: string
}
```
and then we can write sth like: 
```suggestion
          city ? <Component city={city} pathname={pathname} cityCode={cityCode} languageCode={languageCode} /> : null
```
and get rid of all the `if (!city)` in the routes right?

Comment (steffenkleinle) [reply web/src/CityContentSwitcher.tsx:134]: Nope, actually we can't, thats part of the whole PR. By rendering the componentns/routes even if the city is not available, we trigger the loading of the routes data (i.e. events, categories, ...). If we only render the component if the city is available, it will be subsequent requests, not simultaneous ones.

**Review by LeandraH — APPROVED** [review]
Nice! In five tests of this branch vs main with Lighthouse (on a sketchy train internet connection) I got an improvement of 6%.

File: api-client/src/index.ts
Comment (steffenkleinle) [reply api-client/src/index.ts:58]: Hmmm not that sure actually, just historically grown. Since all the things in the api-client should only be used via the exports from index.ts, it should not matter (so also changing it if there is a reason to change it would be okay).

File: web/src/index.ejs
Comment (steffenkleinle) [reply web/src/index.ejs:60]: The sites endpoint is only loaded for the landing page anymore, so I figured it does not make sense to always preload it.

File: web/src/routes/CategoriesPage.tsx
Comment (steffenkleinle) [reply web/src/routes/CategoriesPage.tsx:90]: I don't think so, no. If `parentsLoading` is true, we should just display a loading spinner. If `city` is null, the loading logic is handled in `CityContentSwitcher`. I admit, this is all a little confusing, but it was the only way I could think of to solve the problem of executing the requests simultaneously.

**Comment by SteffiStuffel** [comment]
> Nice! In five tests of this branch vs main with Lighthouse (on a sketchy train internet connection) I got an improvement of 6%.

hmh, i do not get any improvement.
On this branch loading time of city of Augsburg is: 769ms
on main it is: 715ms
Details: WLAN, disable all caching, reload page for several times and calculate the average.

**Comment by SteffiStuffel** [comment]
> @SteffiStuffel we can probably close the epic once this two PRs are merged right? https://issues.tuerantuer.org/browse/IGAPP-1216 Or anything else left to do?

Sure we can close it, when we are happy with the loading times for now.

**Comment by steffenkleinle** [comment]
> > Nice! In five tests of this branch vs main with Lighthouse (on a sketchy train internet connection) I got an improvement of 6%.
> 
> hmh, i do not get any improvement. On this branch loading time of city of Augsburg is: 769ms on main it is: 715ms Details: WLAN, disable all caching, reload page for several times and calculate the average.

TBH I think its not really possible to measure the effect in localhost. I tried both lighthouse and @SteffiStuffel's method to look at the loading times in firefox and it seems to be super dependent on the time executed and it fluctuates quite some even with throttling enabled. Without merging main into my feature branch, I could on average verify @SteffiStuffel's result with an average of 662ms for main and 690ms for testumgebung/de, however, with an updated feature branch it was suddenly 687ms on main and 650ms on the feature branch...

What I can see, however, is that on the feature branch the second request (the children endpoint) is executed always around 1.1s while on the main branch its only around 1.5s. In (my simple) theory my feature branch should therefore be 400ms faster, not sure why it is not showing.

Anyway, my conclusion would be to try this out, look at the times in google search console and if it gets worse/doesn't show an improvement we could revert it. Except @sarahsporck or @f1sh1918 have some other ideas.

**Review by sarahsporck — COMMENTED** [review]
I'm fine with testing it out and see the result. While testing whether this pr actually reduces simultanieous requests I found that these changes lead to more requests allover than the current main branch.
At least for the user story I tested.

For the same actions (navigating to a city, navigating to a leaf category, changing location) this pr makes more requests compared to the main branch. This makes me unsure whether this actually improves loading times. Any opinions?
this pr:
Bildschirmfoto 2023-06-22 um 10 13 50
main:
Bildschirmfoto 2023-06-22 um 10 18 23

**Comment by steffenkleinle** [comment]
> I'm fine with testing it out and see the result. While testing whether this pr actually reduces simultanieous requests I found that these changes lead to more requests allover than the current main branch. At least for the user story I tested.
> 
> For the same actions (navigating to a city, navigating to a leaf category, changing location) this pr makes more requests compared to the main branch. This makes me unsure whether this actually improves loading times. Any opinions?

The main reason why we want to cut the loading times is SEO and that google is rating our loading times as too big. This would not be affected by the use case you tested as google is always just indexing single pages without changing between pages/cities and inside one city, the amount of requests stays exactly the same.

The additional requests in your screenshot are:
- lkaugsburg in the start (not an additional request, on main the sites endpoint is loaded instead which is not on your screenshot)
- sites after going to city change
- aichach-friedberg after selecting that city

Therefore, the only additional requests are related to changing a city which is, at least in my opinion, not a normal or even frequent use case for normal users (except selecting the wrong city) or SEO.

Conclusion: In my opinion, we should merge this if we expect or hope for a loading time improvement while staying inside one city and should ignore the use case of changing cities.

**Review by sarahsporck — APPROVED** [review]
I guess I'm happy with you're explanation. Then let's go and see if there is an improvement :)

**Pull request merged**
